browser/tor/tor_control.cc

```diff
@@ -0,0 +1,1512 @@
+    return false;
+  }
+
+  // Read up to 33 octets.  We should need no more than 32, so 33 will
+  // indicate the file is abnormally large.
+  constexpr size_t kBufSiz = 33;
+  char buf[kBufSiz];
+  int nread = cookiefile.ReadAtCurrentPos(buf, kBufSiz);
+  if (nread < 0) {
+    VLOG(0) << "tor: failed to read Tor control auth cookie";
+    return false;
+  }
+  if (nread > 32) {
+    VLOG(0) << "tor: control auth cookie too large";
+    return false;
+  }
+
+  // Success!
+  cookie.assign(buf, buf + nread);
+  mtime = info.last_accessed;
+  VLOG(3) << "Control cookie " << base::HexEncode(buf, nread)
+             << ", mtime " << mtime;
+  return true;
+}
+
+// EatControlPort(port, mtime)
+//
+//      Try to read the control port number.  Return true and set
+//      port and mtime if successful; return false on failure.
+//
+bool TorControl::EatControlPort(int& port, base::Time& mtime) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(watch_sequence_checker_);
+  DCHECK(polling_);
+
+  // Open the control port file.
+  base::FilePath portpath = watch_dir_path_.AppendASCII("controlport");
+  base::File portfile(portpath, base::File::FLAG_OPEN|base::File::FLAG_READ);
+  if (!portfile.IsValid()) {
+    VLOG(0) << "tor: failed to open control port";
+    return false;
+  }
+
+  // Get the file's info, including modification time.
+  base::File::Info info;
+  if (!portfile.GetInfo(&info)) {
+    VLOG(0) << "tor: failed to stat control port";
+    return false;
+  }
+
+  // Read up to 27/28 octets, the maximum we will ever need.
+#if defined(OS_WIN)
+  const char mintmpl[] = "PORT=1.1.1.1:1\r\n";
+  const char maxtmpl[] = "PORT=255.255.255.255:65535\r\n";
+#else
+  const char mintmpl[] = "PORT=1.1.1.1:1\n";
+  const char maxtmpl[] = "PORT=255.255.255.255:65535\n";
+#endif
+  const size_t kBufSiz = strlen(maxtmpl);
+  char buf[kBufSiz];
+  int nread = portfile.ReadAtCurrentPos(buf, sizeof buf);
+  if (nread < 0) {
+    VLOG(0) << "tor: failed to read control port";
+    return false;
+  }
+  if (static_cast<size_t>(nread) < strlen(mintmpl)) {
+    VLOG(0) << "tor: control port truncated";
+    return false;
+  }
+  DCHECK(static_cast<size_t>(nread) <= sizeof buf);
+
+  std::string text(buf, 0, nread);
+
+  // Sanity-check the content.
+  if (!base::StartsWith(text, "PORT=", base::CompareCase::SENSITIVE) ||
+#if defined(OS_WIN)
+      !base::EndsWith(text, "\r\n", base::CompareCase::SENSITIVE)) {
+#else
+      !base::EndsWith(text, "\n", base::CompareCase::SENSITIVE)) {
+#endif
+    VLOG(0) << "tor: invalid control port: "
+               << "`" << text << ";";  // XXX escape
+    return false;
+  }
+
+  // Verify that it's localhost.
+  const char expected[] = "PORT=127.0.0.1:";
+  if (!base::StartsWith(text, expected, base::CompareCase::SENSITIVE)) {
+    VLOG(0) << "tor: control port has non-local control address";
+    return false;
+  }
+
+  // Parse it!
+#if defined(OS_WIN)
+  std::string portstr(text, strlen(expected), nread - 2 - strlen(expected));
```

Comment:
rather than have all these OS_WIN/else defines, I think it would be better to have one that defines either \r\n or \n and then use that value https://github.com/brave/brave-core/pull/4802/files#diff-dfb7eb2d0039cf56124194955bbfb005R320 and length here instead of 2 and 1.

Reply:
fixed in 12c003763561ef5dc1ef014d4769d24a84c58883